torch/distributed/_shard/checkpoint/storage.py

```diff
@@ -0,0 +1,167 @@
+import abc
+from typing import List, Dict
+
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+
+class StorageWriter(abc.ABC):
+    """
+    Interface used by ``save_state_dict`` to write to storage.
+
+    A subclass should expect the following sequence of calls by ``save_state_dict``
+
+    1) prepare_storage with a dict of storage_key -> size. This can be used to
+        pre-allocate storage.
+    2) write_bytes
+    3) write_tensors.
+    4) (if rank == 0) write_metadata().
+
+    Rank 0 will wait for (2) and (3) before writing metadata.
+    The writes from (2) and (3) are initiated before any waiting is done.
+    """
+
+    @abc.abstractmethod
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        """
+        Initiate writes for all requests in `requests`.
+
+        Writing can happen asynchronously and/or concurrently. A blocking
+        implementation is valid.
+
+        Args:
+            requests (List[BytesWriteRequest]): A list of requests to write
+        Returns:
+            A future that completes once all writes have finished.
+        """
+        pass
+
+    @abc.abstractmethod
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
```

Comment:
The storage layer can write the requests w/ additional customizations (e.g. chunking/batching/quantization etc.) so it would be important for this layer to clarify how storage metadata can be persisted using the APIs provided.

Essentially there are 2 options:

**Option 1:**
change `write_tensors()` to return `List[T]` where `T` is the type of storage metadata (can be defined by `StorageWriter` implementation) and `write_metadata()` takes additional `List[T]` as storage metadata tied to the tensor metadata.

Pros:
a. No cross-rank communication required for `StorageWriter`. The caller is responsible for that if needed (or if something needs to be called in rank 0 only).
b. API layer is explicit about when `StorageWriter` should return the storage metadata.

Cons:
a. Introduction of `T` where `T` is to some extent freeform.

**Option 2:**
Make storage metadata handling an implementation detail inside a writer.

Example pseudo SPMD impl:

```
def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
  storage_metadata = self._do_write(requests)
  self._local_storage_metadata = storage_metadata

def write_metadata(self, metadata: Metadata) -> None:
  storage_metadata_all_ranks = pg.gather(self._local_storage_metadata)
  if pg.my_rank == 0:
    full_metadata = self._transform(metadata, storage_metadata_all_ranks)
    self._do_write(full_metadata)
```

Pros:
1. No `T` introduced

Cons:
1. Storage layer needs to be communication aware so the layering may be unclear (e.g. there may need to be SPMD/non-SPMD version even though the per rank IO logic is exactly the same).
2. Caller contract change (e.g. for SPMD `write_metadata()` needs to be called in every rank not just rank 0)
3. Implicit metadata handling.

Given these, option 1 seems preferred.

Reply:
@xunnanxu  There's a third option worth considering.

Rank0 computes a global plan and scatters it before read/write starts.

Pros:
- No reconciliation at the end.
- No collectives in storage layer

Cons:
- Storage layer needs to be able to plan for non-local rank

I believe Option 1 is better still.

On point (1). I assume `write_tensor` would return `List[StorageMetadata]` which, in turn, would be defined as the following:

```
class StorageMetadata:
  fqn: str #FQN was provided in write request
  storage_data: Any # object that is serializable by `torch.distributed.gather_object`
```

This is follow up work that should happen after this PR lands. I'm tracking it here: https://github.com/pytorch/pytorch/issues/76395

Reply:
The option 3 you mentioned may not work for certain storage layers (or maybe it's me not understanding the plan and you could help clarify). I assume in a perfect world, the state dict saver can be reused. And S3/Azure blob can be plugged in a `StorageWriter` implementations. In that case, a specific storage implementation (say Linux fs) may not have file names but only file descriptors - this implementation clearly has its own quirks that cannot be foreseen by higher level abstractions, and it doesn't make sense to force it to honor "storage keys" from this proposal because it just can't. In that sense the most sensible way would be to let the `StorageWriter` layer determine the actual ids (FDs or whatever) and return those to caller to merge them into the final metadata.